Only add the nonce claim to the ID Token (mapper for backwards compatibility) (#27588)
keycloak/keycloak: docs/documentation/upgrading/topics/keycloak/changes-25_0_0.adoc

= Nonce claim is only added to the ID token

> The nonce claim is now only added to the ID token strictly following the OpenID Connect Core 1.0 specification. The spec says that that this claim is compulsory inside the https://openid.net/specs/openid-connect-core-1_0.html#IDToken[ID token] when the same parameter was sent in the authorization request. Besides it recommends to not add it after a https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse[refresh request]. Previously, the claim was set to all the tokens (Access, Refresh and ID) in all the responses (refresh included). A new `Nonce backwards compatible` mapper is also included in the software that can be assigned to client scopes to revert to the old behavior.

Reviewer: Is it possible to also add another paragraph or sentence that people may need this mapper if they use our `keycloak.js` adapter of the older version than Keycloak 24?

As there is this related bug for keycloak.js https://github.com/keycloak/keycloak/issues/26651 , which caused that JS adapter incorrectly checked the "nonce" in the access token and which was removed in Keycloak 24. So AFAIK keycloak.js applications of older version that Keycloak 24 will still have this check and hence they won't work (unless `Nonce backwards compatible` mapper is added).

Author: Done! Now @andymunro will do his magic. :stuck_out_tongue_closed_eyes:

= Nonce claim is only added to the ID token

> The nonce claim is now only added to the ID token strictly following the OpenID Connect Core 1.0 specification. The spec says that the claim is compulsory inside the https://openid.net/specs/openid-connect-core-1_0.html#IDToken[ID token] when the same parameter was sent in the authorization request. Besides it recommends to not add the `nonce` after a https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse[refresh request]. Previously, the claim was set to all the tokens (Access, Refresh and ID) in all the responses (refresh included).

Reviewer 2: ```suggestion
The nonce claim is now only added to the ID token strictly following the OpenID Connect Core 1.0 specification. As indicated in the specification, the claim is compulsory inside the https://openid.net/specs/openid-connect-core-1_0.html#IDToken[ID token] when the same parameter was sent in the authorization request. The specification also recommends to not add the `nonce` after a https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse[refresh request]. Previously, the claim was set to all the tokens (Access, Refresh and ID) in all the responses (refresh included).
```

Author: Thanks! Done!

> A new `Nonce backwards compatible` mapper is also included in the software that can be assigned to client scopes to revert to the old behavior. For example, the JS adapter checked the returned `nonce` claim in all the tokens before fixing issue https://github.com/keycloak/keycloak/issues/26651[#26651] in version 24.0.0. Therefore, if an old version of the JS adapter is used, the mapper should be added to the required clients via client scopes.

Reviewer 2: ```suggestion
A new `Nonce backwards compatible` mapper is also included in the software that can be assigned to client scopes to revert to the old behavior. For example, the JS adapter checked the returned `nonce` claim in all the tokens before fixing issue https://github.com/keycloak/keycloak/issues/26651[#26651] in version 24.0.0. Therefore, if an old version of the JS adapter is used, the mapper should be added to the required clients by using client scopes.
```

Author: Thanks! Done!
